core/server/master/src/main/java/alluxio/master/file/AccessTimeUpdater.java

```diff
@@ -0,0 +1,171 @@
+        ServerConfiguration.getMs(PropertyKey.MASTER_FILE_ACCESS_TIME_UPDATER_SHUTDOWN_TIMEOUT));
+  }
+
+  /**
+   * Constructs a new {@link AccessTimeUpdater} with time configurations.
+   */
+  @VisibleForTesting
+  public AccessTimeUpdater(FileSystemMaster fileSystemMaster, InodeTree inodeTree,
+      JournalSystem journalSystem, long flushInterval, long updatePrecision, long shutdownTimeout) {
+    mFileSystemMaster = fileSystemMaster;
+    mInodeTree = inodeTree;
+    mAccessTimeUpdates = new ConcurrentHashMap<>();
+    mFlushInterval = flushInterval;
+    mUpdatePrecision = updatePrecision;
+    mShutdownTimeout = shutdownTimeout;
+    journalSystem.addJournalSink(mFileSystemMaster, this);
+  }
+
+  public void start() {
+    start(mFlushInterval > 0 ? Executors.newSingleThreadScheduledExecutor(
+        ThreadFactoryUtils.build("AccessTimeUpdater-%d", true)) : null);
+  }
+
+  @VisibleForTesting
+  public synchronized void start(ScheduledExecutorService executorService) {
+    if (mExecutorService != null && mExecutorService != executorService
+        && !mExecutorService.isShutdown()) {
+      stop();
+    }
+    mExecutorService = executorService;
+  }
+
+  @Override
+  public void beforeShutdown() {
+    if (mExecutorService != null) {
+      flushUpdates();
+    }
+  }
+
+  public synchronized void stop() {
+    if (mExecutorService != null) {
+      ThreadUtils.shutdownAndAwaitTermination(mExecutorService, mShutdownTimeout);
+    }
+  }
+
+  /**
+   * Update inode last access time. Requires at least read lock acquired for the inode.
+   *
+   * @param context the journal context
+   * @param inode the target inode
+   * @param opTimeMs the last access time to set on the inode
+   */
+  public void updateAccessTime(JournalContext context, Inode inode, long opTimeMs) {
+    if (opTimeMs - inode.getLastAccessTimeMs() > mUpdatePrecision) {
+      try (LockResource lr = mInodeTree.getInodeLockManager().lockUpdate(inode.getId())) {
+        if (mExecutorService != null) {
+          // journal update asynchronously
+          UpdateInodeEntry entry = mInodeTree.updateInodeAccessTimeNoJournal(inode.getId(),
+              opTimeMs);
+          scheduleJournalUpdate(entry);
+        } else {
+          mInodeTree.updateInode(context, UpdateInodeEntry.newBuilder()
+                  .setId(inode.getId())
+                  .setLastAccessTimeMs(opTimeMs)
+                  .build());
+        }
+      }
+    }
+  }
+
+  private void scheduleJournalUpdate(UpdateInodeEntry entry) {
+    mAccessTimeUpdates.put(entry.getId(), entry.getLastAccessTimeMs());
+    if (mUpdateScheduled.compareAndSet(false, true)) {
+      mExecutorService.schedule(this::flushScheduledUpdates, mFlushInterval, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private void journalAccessTime(JournalContext context, UpdateInodeEntry entry) {
+    context.append(Journal.JournalEntry.newBuilder().setUpdateInode(entry)
```

Comment:
I think we can just inline this at the call site.

Reply:
Inlined.